federationsender/internal/api.go

```diff
@@ -38,3 +44,98 @@ func NewFederationSenderInternalAPI(
 		queues:     queues,
 	}
 }
+
+func (a *FederationSenderInternalAPI) isBlacklistedOrBackingOff(s gomatrixserverlib.ServerName) (*statistics.ServerStatistics, error) {
+	stats := a.statistics.ForServer(s)
+	if stats.Blacklisted() {
+		return stats, &api.FederationClientError{
+			Blacklisted: true,
+		}
+	}
+	// Call BackoffIfRequired with a closed channel to make it return immediately.
+	// It will return the duration to backoff for.
+	var duration time.Duration
+	interrupt := make(chan bool)
+	close(interrupt)
+	var bo atomic.Bool
+	duration, _ = stats.BackoffIfRequired(bo, interrupt)
```

Comment:
At the moment, this call will affect the backoff intervals for the destination queue as it will increase the `backoffCount`. This will mean that the next time the destination queue calls this, it'll think that it has backed off more times than it really has and so the interval will go right up (possibly leading us to blacklist early).

The simple way around this is to make an amendment to `BackoffIfRequired` so that if the interrupt channel is already closed, we stop there before increasing the backoff count, e.g.
```
func (s *ServerStatistics) BackoffIfRequired(backingOff atomic.Bool, interrupt <-chan bool) (time.Duration, bool) {
	if started := s.backoffStarted.Load(); !started {
		return 0, false
	}

	// Check if the interrupt channel is already closed. If it is
	// then this call should have no side effects but still return
	// the current duration.
	select {
	case <-interrupt:
		count := s.backoffCount.Load()
		return time.Second * time.Duration(math.Exp2(float64(count))), true
	default:
	}

...
```

Reply:
Thanks for this!

Reply:
This doesn't seem to have the intended behaviour I want. A sytest times out because we never actually ever mark the backoff as complete so always return "backoff for 1s" with this code...